### PR TITLE
Delta Lake: Improve error message related to storage layers which allow write collisions

### LINKED ISSUE #14464: Improve error message on DML operations when `delta.enable-non-concurrent-writes` is `false`
```
io.trino.spi.TrinoException: Inserts are not supported on the s3 filesystem
at io.trino.plugin.deltalake.DeltaLakeMetadata.beginInsert(DeltaLakeMetadata.java:1247)
```

When `delta.enable-non-concurrent-writes` is set to `false`, the error message thrown is rather misleading.

Stating that the writes are not supported is actually incorrect.

The error message should point to the setting `delta.enable-non-concurrent-writes` as a way to enable the writes even though limited.

### PR DESCRIPTION
**Pull request by findinpath**
## Description

Fixes #14464

## Release notes

(x) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

**Comment by alexjo2144** [comment]
This seems like something that should go in the docs, not in an error message. An individual user doesn't have the ability to turn this on, like they do with session properties. 

It should also come along with a warning that writes are safe as long as you only have one cluster doing them.

**Comment by findinpath** [comment]
@alexjo2144 what about using this error message:

`Inserts are not enabled on the %s filesystem.`

?

From a usability point of view, we should give somehow the hint in the message which can be used by sysadmin to adapt the connector configuration if needed.

**Comment by mdesmet** [comment]
> It should also come along with a warning that writes are safe as long as you only have one cluster doing them.

According to my understanding when competing writes are happening, the order of execution is random and therefore also the result, even when only one cluster is used. So enabling should probably come with a warning in any case.

I personally think it is user-friendly to also mention the property to set. We do the same in https://github.com/trinodb/trino/blob/ea0255906f827ccb8332f5f2afc155b21a2f64a8/plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java#L693-L696

**Comment by hashhar** [comment]
> According to my understanding when competing writes are happening, the order of execution is random and therefore also the result, even when only one cluster is used.

That's how transaction failures work on most systems with serializable isolation level. If multiple queries hold a table lock only one of them will win but which one isn't guaranteed. The failed one has to be retried.

You don't notice it as often with other databases since the default isolation is generally never serializable.

**Comment by alexjo2144** [comment]
> what about using this error message:
Inserts are not enabled on the %s filesystem.

This is better, to me.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 1]
```diff
@@ -1257,7 +1257,10 @@ public ConnectorInsertTableHandle beginInsert(ConnectorSession session, Connecto
         DeltaLakeTableHandle table = (DeltaLakeTableHandle) tableHandle;
         if (!allowWrite(session, table)) {
             String fileSystem = new Path(table.getLocation()).toUri().getScheme();
-            throw new TrinoException(NOT_SUPPORTED, format("Inserts are not supported on the %s filesystem", fileSystem));
+            throw new TrinoException(
+                    NOT_SUPPORTED,
+                    format("Inserts are not enabled on the %1$s filesystem in order to avoid eventual data corruption which may be caused by concurrent data modifications on the table. " +
+                            "Writes to the %1$s filesystem can be however enabled with the 'delta.enable-non-concurrent-writes' configuration property.", fileSystem));
```
Comment (findepi) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:1263]: Define constant for `delta.enable-non-concurrent-writes`

**Pull request merged**